### ISSUE TITLE
What version of the original GNN does this code implement?

### ISSUE THREAD

**Issue by danielegrattarola** [mtiezzi/torch_gnn#1]
Hi, 

sorry if this is not the best place to ask but I thought it would be easier to reference code here. 

I am trying to re-implement the original GNN model ([Scarselli et al., 2009, TNNLS](https://ieeexplore.ieee.org/document/4700287)) and I'm having trouble matching your code to the description in the paper. 

From what I understand, these lines:

https://github.com/mtiezzi/torch_gnn/blob/81d014fcf409808b5a6939fca6cacfdf13d82cd8/net.py#L48-L66

implement the version of the model called "nonlinear (nonpositional) GNN" in the paper (Section 2.D, point 2), where it is said that "h_w is realized by a multilayered FNN."

If that is the case, then I could not find the implementation of the additional loss term that should ensure that the operator is contractive. I'm likely missing something, so I would really appreciate if you could point me in the right direction. 

I'm mostly interested in how to make the operator contractive, so even if you have a different solution I'd be really interested. 

Thanks, 
Daniele

**Comment by mtiezzi** [comment]
Hello Daniele,
regarding the question on the model, yes this is the implementation of the nonpositional GNN exploiting the arc-wise h_w state update function.

We developed the code under the supervision of the authors of the original model (Franco Scarselli) in Tensorflow [(here the original repo).](https://github.com/mtiezzi/gnn) and in PyTorch (this repo).

We decided all togheter  to relax the requirements of the state transition function being a contractive map (hence, we do not need to enforce the contractive condition) because we have seen that the model is able to undergo the converge procedure even without this costly requirement. 

As you can see, we stop the convergence procedure once all the states reach convergence, or otherwise we stop at a certain iteration (default=50).

Have also a look at the [documentation page](https://mtiezzi.github.io/gnn_site/GNN/model.html#introduction)

If you want more information on the contractive operator version, you can look at the old [MATLAB implementation](https://www3.diism.unisi.it/~franco/Research/GNNDownload.php).

Feel free to use my implementation as a baseline :)

**Comment by danielegrattarola** [comment]
Thanks for the quick reply! I will try to see if my version converges as well without the contraction requirement.

Also, thanks for publishing this code. It has been very useful!

Cheers, 
Daniele

**Comment by mtiezzi** [comment]
You're welcome!
Have a look also at our [Lagrangian Propagation GNNs](https://github.com/mtiezzi/lpgnn) (we used the original GNN as a baseline there).
Best,
Matteo
